wit/types.wit

```diff
@@ -0,0 +1,158 @@
+// The `wasi:http/types` interface is meant to be imported by components to
+// define the HTTP resource types and operations used by the component's
+// imported and exported interfaces.
+default interface types {
+  use io.streams.{input-stream, output-stream}
+  use poll.poll.{pollable}
+  
+  // This type corresponds to HTTP standard Methods.
+  variant method {
+    get,
+    head,
+    post,
+    put,
+    delete,
+    connect,
+    options,
+    trace,
+    patch,
+    other(string)
+  }
+
+  // This type corresponds to HTTP standard Related Schemes.
+  variant scheme {
+    HTTP,
+    HTTPS,
+    other(string)
+  }
+
+  // This type enumerates the different kinds of errors that may occur when
+  // initially returning a response.
+  variant error {
+      invalid-url(string),
+      timeout-error(string),
+      protocol-error(string),
+      status-error(u16),
+      unexpected-error(string)
+  }
```

Comment:
This is kind of weird enum, since it mixes status codes with non-HTTP errors. What's the intent here?

Also, the list is pretty limited. Perhaps we should include errors from the `Proxy-Status` RFC (i.e. https://www.iana.org/assignments/http-proxy-status/http-proxy-status.xhtml)?

Reply:
Good question; I mostly just lifted this from one of the other proposals.  Aligning with HTTP semantically-defined errors sounds generally good, but don't we already get that from the `status-code`?  Perhaps this is something best hashed out in a separate issue.  For now, I've added a [TODO](https://github.com/WebAssembly/wasi-http/pull/3/commits/0bd4751db8c8858f964ecb37dac6cf61bbd82205) to the type definition.

Reply:
I think we should get rid of `status-error` in general. From the client perspective, if HTTP communication worked, that's not an error, obviously the client surfaces the received status code up to the caller, but it's up to the caller to decide if it is an error.

A classic example of this is whether `404` is an error or not. It really depends on the calling context, and if you force a programmer to do something special when they actually expect a `404` then it becomes problematic.

Reply:
That makes sense to me.

Reply:
[Fixed](https://github.com/WebAssembly/wasi-http/pull/3/commits/5bac3093efaea7eaacee3f79d6fcf79aad68270d)

Reply:
Filed #5 for the more general question.